### PR TITLE
Add dict comprehension

### LINKED ISSUE #41448: [JIT] Support Dict comprehension
Currently, dict comprehension is not supported by JIT but it is widely used in python codes. Is there any plans to support it?

```
class MyModule(nn.Module):
    def forward(self):
        x = {i: i for i in range(2)}
        return x

model = MyModule()
model_s = torch.jit.script(model)
```
we will get:
```
torch.jit.frontend.UnsupportedNodeError: DictComp aren't supported:
  File "main_test.py", line 6
    def forward(self):
        x = {i: i for i in range(2)}
            ~ <--- HERE
        return x
```
we may use a `for-loop` to gradually add elements to the dict as a workaround, but it's a pretty bad user experience.

cc @suo @gmagogsfm

### PR DESCRIPTION
**Pull request by ansley**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#47774 Add dict comprehension**

Differential Revision: [D25615464](https://our.internmc.facebook.com/intern/diff/D25615464)

### CODE REVIEW

File: torch/csrc/jit/frontend/parser.cpp [hunk 1]
```diff
@@ -144,6 +144,14 @@ struct ParserImpl {
       } break;
       case '{': {
         L.next();
+        // If we have a dict literal, `keys` and `values` will store the keys
+        // and values used in the object's construction. EDGE CASE: We have a
+        // dict comprehension, so we'll get the first element of the dict
```
Comment (gmagogsfm) [inline torch/csrc/jit/frontend/parser.cpp:149]: What is "first element of the dict comprehension"? Jus the first key?

File: aten/src/ATen/core/interned_strings.h [hunk 2]
```diff
@@ -56,7 +56,7 @@ namespace c10 {
   _(prim, ReturnStmt)                \
   _(prim, BreakStmt)                 \
   _(prim, ContinueStmt)              \
-  _(prim, ListComprehensionScope)    \
+  _(prim, ComprehensionScope)        \
```
Comment (gmagogsfm) [inline aten/src/ATen/core/interned_strings.h:59]: I think this breaks BC test, also might have ramification for mobile. Could we still keep `ListComprehensionScope`?

**Review by gmagogsfm — COMMENTED** [review]
Overall looks good, just a few comments. Also there seem to be some broken tests, could you take care of them in next revision?

File: aten/src/ATen/core/interned_strings.h (hunk 2 above)
Comment (eellison) [reply aten/src/ATen/core/interned_strings.h]: `ListComprehension` isn't emitted as  an operator, it's a node place holder, just like prim::If, that exists temporarily in initial compilation. Changing the name doesn't effect BC (michael recently did it)

File: torch/csrc/jit/frontend/ir_emitter.cpp [hunk 3]
```diff
@@ -1299,6 +1299,54 @@ struct to_ir {
...
+    Node* n =
+        graph->insertNode(create(prim::ListComprehensionScope, dc.range(), 0));
+    auto* comprehension_block = n->addBlock();
+    pushFrame(comprehension_block);
+    WithInsertPoint guard(comprehension_block);
+    auto emit_body = [&]() {
+      auto k = emitExpr(dc.key());
+      auto v = emitExpr(dc.value());
+      if (!type_set) {
+        dict_value->setType(DictType::create(k->type(), v->type()));
```
[the remark this answers is not in the document]
Comment (ansley) [reply torch/csrc/jit/frontend/ir_emitter.cpp:1336]: I don't believe that there are any possible paths in the code that could lead us to a dict comprehension with heterogenous key/value types. For example, the following code will raise an error before the dict comprehension is even created.

```python
@torch.jit.script
def fn():
    l = [("s", 1), (2, "f"), (1.0, 3)]
    return {k : v for k, v in l}
```

File: torch/csrc/jit/frontend/ir_emitter.cpp [hunk 4]
```diff
@@ -1299,6 +1299,54 @@ struct to_ir {
...
+    const auto loc = dc.range();
+    const auto targets_list = List<Expr>::create(dc.range(), {dc.target()});
+    const auto itrs = List<Expr>::create(dc.range(), {dc.iter()});
+
+    // If 1) there is no type hint, and 2) we emit this Value over an unrolled
+    // iterable of length 0, then we emit a Dict[Tensor, Tensor]
+    Value* dict_value =
+        graph->insertNode(graph->create(prim::DictConstruct, 1))
+            ->output()
+            ->setType(DictType::create(TensorType::get(), TensorType::get()));
```
[the remark this answers is not in the document]
Comment (gmagogsfm) [reply torch/csrc/jit/frontend/ir_emitter.cpp:1314]: Is this `setType` necessary?

**Review by gmagogsfm — APPROVED** [review]
Looks good, just a minor comment

File: torch/csrc/jit/frontend/ir_emitter.cpp [hunk 5]
```diff
@@ -1301,6 +1301,50 @@ struct to_ir {
     return list_value;
   }
 
+  Value* emitDictComprehension(const DictComp& dc, const TypePtr& type_hint) {
+    const auto loc = dc.range();
+    const auto targets_list = List<Expr>::create(dc.range(), {dc.target()});
+    const auto itrs = List<Expr>::create(dc.range(), {dc.iter()});
+
+    Value* dict_value =
+        graph->insertNode(graph->create(prim::DictConstruct, 1))->output();
```
Comment (eellison) [inline torch/csrc/jit/frontend/ir_emitter.cpp:1310]: The "unrolled loop" thing is for heterogenous types which we want to allow to be iterated over. 
For example: 
```
for elem in (1, 1.5, "abc"):
    print(elem)
```
It also applies with ModuleLists, which contain different Modules, which are different types.
```[module(x) for module in self.module_list]```. You could have an empty tuple  or an empty modulelist, so what would the output type be?  `[elem for elem in tuple()]`. We could either  throw or give the same default as an  empty list constructor. For lists, we set it as the type of empty list constructor `x = []`. I think we should do  the same in dict comprehensions. 

All that entails is setting the DictConstruct initial type here to be [Str, Tensor] (same `x = {}`).

**Review by eellison — APPROVED** [review]
Looks great, one small comment

**Pull request closed without merging**